compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5253,8 +5253,13 @@ private SymbolEnv handleFilterClauses (BLangExpression filterExpression) {
             dlog.error(filterExpression.pos, DiagnosticErrorCode.INCOMPATIBLE_TYPES,
                     symTable.booleanType, actualType);
         }
-        SymbolEnv filterEnv = typeNarrower.evaluateTruth(filterExpression, selectClauses.peek(), queryEnvs.pop());
-        queryEnvs.push(filterEnv);
+        SymbolEnv filterEnv;
+        if (selectClauses.peek() != null) {
+           filterEnv = typeNarrower.evaluateTruth(filterExpression, selectClauses.peek(), queryEnvs.pop());
+           queryEnvs.push(filterEnv);
+        } else {
+            filterEnv = queryEnvs.peek();
```

Comment:
Instead of peeking prior env, type narrow to do clause

Reply:
yeah. figured that out :)